rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/util/CovidRDTJsonFormUtils.java

```diff
@@ -147,4 +163,40 @@ private String getLoggedInUserPreferredName() {
         final AllSharedPreferences allSharedPreference = RDTApplication.getInstance().getContext().allSharedPreferences();
         return allSharedPreference.getANMPreferredName(allSharedPreference.fetchRegisteredANM());
     }
+
+    private JSONArray getLocations() throws JSONException {
+        JSONArray jsonArray = new JSONArray();
+        AllSharedPreferences allSharedPreferences = RDTApplication.getInstance().getContext().allSharedPreferences();
+        String defaultLocationUuid = allSharedPreferences.fetchDefaultLocalityId(allSharedPreferences.fetchRegisteredANM());
+        LinkedHashMap<String, TreeNode<String, Location>> locationMap = LocationHelper.getInstance().map();
+        List<String> locations = filterLocations(defaultLocationUuid, locationMap);
+        for (String location : locations) {
+            JSONObject option = new JSONObject();
+            option.put(JsonFormConstants.KEY, location);
+            option.put(JsonFormConstants.TEXT, location);
+            option.put(JsonFormConstants.OPENMRS_ENTITY, "");
+            option.put(JsonFormConstants.OPENMRS_ENTITY_ID, "");
+            jsonArray.put(option);
+        }
+        return jsonArray;
+    }
+
+    private List<String> filterLocations(String locationId, LinkedHashMap<String, TreeNode<String, Location>> map) {
+        Map.Entry<String, TreeNode<String, Location>> entry = map.entrySet().iterator().next();
+        if (entry != null) {
+            if (entry.getKey().equals(locationId)) {
+                List<String> locations = new ArrayList<>();
+                for (Map.Entry<String, TreeNode<String, Location>> childEntry : map.entrySet()) {
+                    locations.add(childEntry.getValue().getLabel());
+                }
+                return locations;
+            }
+            else {
+                return filterLocations(locationId, entry.getValue().getChildren());
+            }
+        }
+        else {
+            return new ArrayList<>();
+        }
```

Comment:
This logic can be simplified a bit by removing the external if/else block.

Reply:
fixed.

Reply:
Also, we may not need this recursion if we start from the immediate parent location.